armoryengine/WalletUtils.py

```diff
@@ -311,3 +310,156 @@ def getBalances(self):
+
+      self.displayName = self.walletId if self.walletId else os.path.splitext(self.filename)[0]
+      self.watchingOnly = entry.watchingOnly
+
+   @property
+   def filename(self):
+      """Extract filename from path on demand."""
+      return os.path.basename(self.path) if self.path else ''
+
+################################################################################
+## WalletList Class - Handles all wallet data fetching and preparation
+################################################################################
+
+class WalletList(object):
+   """
+   Handles wallet data fetching and formatting for UI rendering.
+   NO GUI code - pure data processing only.
+   """
+
+   def __init__(self):
+      self.wallets = []
+
+   def update(self):
+      """Fetch and process wallet list, updating internal state."""
+      try:
+         rawWalletData = TheBridge.wltManager.listWallets()
+      except Exception as e:
+         LOGERROR(f"Failed to fetch wallet list: {e}")
+         self.wallets = []
+         return
+
+      # Simple check-and-replace: iterate entries, keep best per walletId
+      self.wallets = []
+      encryptedWallets = []
+      bestByWalletId = {}
+
+      for entry in rawWalletData:
+         stateName = str(entry.state).lower()
+
+         # Encrypted wallets: walletId is unreadable, show ALL
+         if stateName == 'encrypted':
+            encryptedWallets.append(WalletListEntry(entry))
+            continue
+
+         walletId = entry.walletId
+         if not walletId:
+            continue
+
+         # Check if this entry should replace existing one
+         if walletId not in bestByWalletId:
+            bestByWalletId[walletId] = entry
+         else:
+            existingState = str(bestByWalletId[walletId].state).lower()
+            if self.replaceEntry(existingState, stateName, bestByWalletId[walletId], entry):
+               bestByWalletId[walletId] = entry
+
+      # Add all encrypted wallets first
+      self.wallets.extend(encryptedWallets)
+
+      # Add best entry per walletId
+      for walletId, entry in bestByWalletId.items():
+         self.wallets.append(WalletListEntry(entry))
+
+   def getFilteredList(self):
+      """Return the filtered wallet list for UI rendering."""
+      return self.wallets
+
+   def replaceEntry(self, existingState, newState, existingEntry, newEntry):
+      """
+      Check if newState should replace existingState.
+      Priority: ready (full) > ready (watch-only) > legacy > unknown
+
+      Note: migrated/loaded states are excluded by listWallets() and never reach here.
+      """
+      # Ready replaces legacy or unknown
+      if newState == 'ready':
+         if existingState != 'ready':
+            return True
+         # Both ready: full wallet replaces watch-only wallet
+         existingIsWO = existingEntry.watchingOnly
+         newIsWO = newEntry.watchingOnly
+         return not newIsWO and existingIsWO
```

Comment:
You do not need to check both states here. Either new wallet is WO and you ignore it, or existing is not WO and you keep it.